Lib/shutil.py

```diff
@@ -1462,10 +1462,10 @@ def which(cmd, mode=os.F_OK | os.X_OK, path=None):
     # If we're given a path with a directory part, look it up directly rather
     # than referring to PATH directories. This includes checking relative to the
     # current directory, e.g. ./script
+    # If not, later logic will further verify.
```

Comment:
I think it's simpler to use `dirname, cmd = os.path.split(cmd)`. If `dirname` isn't empty, use `path = [dirname]`, else use the supplied `path` or the environment's `PATH` value. For example:

```python
    use_bytes = isinstance(cmd, bytes)

    # If we're given a path with a directory part, look it up directly rather
    # than referring to PATH directories. This includes checking relative to
    # the current directory, e.g. ./script
    dirname, cmd = os.path.split(cmd)
    if dirname:
        path = [dirname]
    else:
        if path is None:
            path = os.environ.get("PATH", None)
            if path is None:
                try:
                    path = os.confstr("CS_PATH")
                except (AttributeError, ValueError):
                    # os.confstr() or CS_PATH is not available
                    path = os.defpath
            # bpo-35755: Don't use os.defpath if the PATH environment variable
            # is set to an empty string

        # PATH='' doesn't match, whereas PATH=':' looks in the current
        # directory
        if not path:
            return None

        if use_bytes:
            path = os.fsencode(path)
            path = path.split(os.fsencode(os.pathsep))
        else:
            path = os.fsdecode(path)
            path = path.split(os.pathsep)

        if sys.platform == "win32":
            # The current directory takes precedence on Windows.
            # FIXME: the current directory should only be included if
            # WinAPI NeedCurrentDirectoryForExePathW(cmd) is true.
            curdir = os.curdir
            if use_bytes:
                curdir = os.fsencode(curdir)
            if curdir not in path:
                path.insert(0, curdir)

    if sys.platform == "win32":
        # PATHEXT is necessary to check on Windows.
        pathext_source = os.getenv("PATHEXT") or _WIN_DEFAULT_PATHEXT
        pathext = [ext for ext in pathext_source.split(os.pathsep) if ext]

        if use_bytes:
            pathext = [os.fsencode(ext) for ext in pathext]
        # See if the given file matches any of the expected path extensions.
        # This will allow us to short circuit when given "python.exe".
        # If it does match, only test that one, otherwise we have to try
        # others.
        # FIXME: the exact name should be checked as well as the name plus
        # each PATHEXT extension, as implemented by both CMD and PowerShell.
        if any(cmd.lower().endswith(ext.lower()) for ext in pathext):
            files = [cmd]
        else:
            files = [cmd + ext for ext in pathext]
    else:
        # On other platforms you don't have things like PATHEXT to tell you
        # what file suffixes are executable, so just pass on cmd as-is.
        files = [cmd]
```

Reply:
@eryksun i've updated to take this into account.. also fixing the FIXMEs while here. Please check again.